File: typescript/pages/CheckoutPage.ts
import { Page, Locator, expect, test } from '@playwright/test';

export class CheckoutPage {
    readonly firstNameField: Locator;
    readonly lastNameField: Locator;
    readonly postalCodeField: Locator;
    readonly continueButton: Locator;
    readonly finishButton: Locator;

    constructor(private page: Page) {
        this.firstNameField = page.getByPlaceholder('First Name');
        this.lastNameField = page.getByPlaceholder('Last Name');
        this.postalCodeField = page.getByPlaceholder('Zip/Postal Code');

        this.continueButton = page.getByTestId('continue');
        this.finishButton = page.getByTestId('finish');
    }

    async enterInformation(firstName: string, lastName: string, postalCode: string) {
        await test.step('Enter checkout information', async () => {
            await this.firstNameField.fill(firstName);
            await this.lastNameField.fill(lastName);
            await this.postalCodeField.fill(postalCode);
            await this.continueButton.click();
        });
    }

    async verifyCheckoutURL(url: string) {
        await test.step('Verify checkout URL', async () => {
            await expect(this.page).toHaveURL(url);
        });
    }

    async verifyTotalIsDisplayed() {
        await test.step('Verify total is displayed', async () => {
            await expect(
                this.page.locator('[data-test="total-label"]'))
                .toBeVisible();
        });
    }

    async clickFinish() {
        await test.step('Click finish button', async () => {
            await this.finishButton.click();
        });
    }

    async verifyCheckoutComplete() {
        await test.step('Verify checkout is complete', async () => {
            await expect(
                this.page
                    .locator('[data-test="checkout-complete-container"]'))
                .toBeVisible();
        });
    }
}

File: python/pages/checkout_page.py
from playwright.sync_api import Page, expect
import allure


class CheckoutPage:

    def __init__(self, page: Page):
        self.page = page
        self.first_name = page.locator('[data-test="firstName"]')
        self.last_name = page.locator('[data-test="lastName"]')
        self.postal_code = page.locator('[data-test="postalCode"]')
        self.continue_button = page.get_by_role(
            "button", name="Continue"
        )

    def verify_checkout_page(self):
        with allure.step("Verify checkout page is displayed"):
            expect(self.page).to_have_url(
                "https://www.saucedemo.com/checkout-step-one.html"
            )

    def enter_checkout_information(
        self,
        first_name: str,
        last_name: str,
        postal_code: str
    ):
        with allure.step("Enter checkout information"):
            self.first_name.fill(first_name)
            self.last_name.fill(last_name)
            self.postal_code.fill(postal_code)

    def continue_checkout(self):
        with allure.step("Continue checkout"):
            self.continue_button.click()

    def verify_checkout_overview_page(self):
        with allure.step("Verify checkout overview page"):
            expect(self.page).to_have_url(
                "https://www.saucedemo.com/checkout-step-two.html"
        )

    def finish_checkout(self):
        with allure.step("Finish checkout"):
            self.page.get_by_role(
                "button", name="Finish"
            ).click()

    def verify_checkout_complete_page(self):
        with allure.step("Verify checkout complete page"):
            expect(self.page).to_have_url(
            "https://www.saucedemo.com/checkout-complete.html"
        )